features/api-layer/src/main/java/org/opennms/features/apilayer/dao/AlarmFeedbackDaoImpl.java

```diff
@@ -0,0 +1,148 @@
+                    .build();
+
+    /**
+     * The collection of listeners interested in alarm feedback, populated via runtime binding.
+     */
+    private final Collection<AlarmFeedbackListener> alarmFeedbackListeners = new CopyOnWriteArrayList<>();
+
+    /**
+     * Add listeners to {@link #alarmFeedbackListeners} during runtime as they become available.
+     */
+    public synchronized void onBind(AlarmFeedbackListener alarmFeedbackListener, Map properties) {
+        LOG.debug("bind called with {}: {}", alarmFeedbackListener, properties);
+
+        if (alarmFeedbackListener != null) {
+            alarmFeedbackListeners.add(alarmFeedbackListener);
+        }
+    }
+
+    /**
+     * Remove listeners from {@link #alarmFeedbackListeners} during runtime as they become unavailable.
+     */
+    public synchronized void onUnbind(AlarmFeedbackListener alarmFeedbackListener, Map properties) {
+        LOG.debug("Unbind called with {}: {}", alarmFeedbackListener, properties);
+
+        if (alarmFeedbackListener != null) {
+            alarmFeedbackListeners.remove(alarmFeedbackListener);
+        }
+    }
+
+    /**
+     * Gets a reference to the {@link FeedbackRepository feedback repository} if it can be found.
+     *
+     * @return the feedback repository or null if none found
+     */
+    private FeedbackRepository getRepository() {
+        return SERVICE_LOOKUP.lookup(FeedbackRepository.class, null);
+    }
+
+    @Override
+    public List<AlarmFeedback> getFeedback() {
+        FeedbackRepository feedbackRepository = getRepository();
+
+        // If no feedback repository is available we will just return an empty result
+        if (feedbackRepository == null) {
+            return Collections.emptyList();
+        }
+
+        try {
+            return feedbackRepository.getAllFeedback()
+                    .stream()
+                    .map(ModelMappers::toFeedback)
+                    .collect(Collectors.toList());
+        } catch (FeedbackException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Override
+    public void submitFeedback(List<AlarmFeedback> alarmFeedback) {
+        FeedbackRepository feedbackRepository = getRepository();
+
+        // If no feedback repository is available that is an error case as we are unable to complete this request
+        if (feedbackRepository == null) {
+            throw new RuntimeException("Could not find a feedback repository");
+        }
+
+        List<org.opennms.features.situationfeedback.api.AlarmFeedback> mappedFeedback = alarmFeedback.stream()
+                .map(ModelMappers::fromFeedback)
+                .collect(Collectors.toList());
+
+        try {
+            feedbackRepository.persist(mappedFeedback);
+        } catch (Exception e) {
+            throw new RuntimeException("Failed to execute query: " + e.getMessage(), e);
+        }
+
+        alarmFeedbackListeners.forEach(listener -> {
```

Comment:
Can we move the logic of invoking the listeners into the repository rather than duplicating it here? As is stand, everyone that wants to use the repository is now responsible for also managing and calling the listeners.